### DISCUSSION TITLE
Flashdrive gets often corrupted on windows

### DISCUSSION THREAD

**Discussion by rootfrogsMakes** [micropython/micropython#12389] [STM32 / Pyboard]
Hello,
Im using the STM32F4 Discoveryboard to explore Micropython.
I love the feature, that I get a PYBFLASH Drive on Windows if I connect the board via USB. Its very nice just to drag & drop the Files to the board.
After a file transfer, I have to restart the board (press Reset Button oder unplug/plug the board) 
Unfortunately the flash drive gets often corrupted if I unplug (or reset) the board without manually click "remove drive" in windows before.
(Although no file transfer takes place when I unplug/reset the board)
Round about each 3rd time I corrupt the flash by unpluging/resetting it without clicking "remove drive" before.

1) Is this normal, or maybe a bug in the Port implementation?
2) Is there any chance to make this more robust? 

I don't understand what the problem is. If no file transfer is active at the time I unplug/reset the board, why does this corrupt my flash drive?

**Comment by robert-hh** [comment]
That is a often reported issue with a PC and the Flash file system at the board, not limited to Windows. The PC OS assumes that it is the only instance writing to the PYBFLASH file system. I has caches, and it writes to the PYBFLASH at it's own discretion, latest when the PYBFLASH is ejected. So there are two error situations.

- The PC has modified the file system, but did not write back the changes when PYBFLASH is disconnected without ejecting. Then changes to the flash may be incomplete.
- MicroPyton at the board has written to the file system, which is not noticed by the PC. If then the PC writes, it will create conflicting changes. For this reason, e.g. CircuitPython sets the internal file access to read-only, as long an the board is mounted to the PC. A similar mechanism is discussed for Micropython, but without any conclusion.

In order to avoid such a trouble, many people do not use the PYBFLASH mechanism at all and prefer using e.g. mpremote  for file transfer.

**Comment by peterhinch** [comment]
@rootfrogsMakes See [my unofficial FAQ](https://github.com/peterhinch/micropython-samples/blob/master/README.md#15-a-usb-pitfall).

**Comment by jimmo** [comment]
As Robert and Peter have suggested, you will have a much better time using `mpremote`.

Just for completeness though, often the cause of issues is not waiting for the writes to be flushed. You will notice that an LED is illuminated when you write to the device over USB. You must wait for this LED to turn off before disconnecting.

**Comment by rootfrogsMakes** [comment]
Thanks for the answers.
I will use mpremote or similar tools in future to transfer the files.
Thanks a lot for the hint.
